src/psyclone/tests/psyir/nodes/extract_node_test.py

```diff
@@ -265,16 +279,219 @@ def test_extract_node_gen():
+    '''
+
+    # Test if there is no clash that the specified postfix is returned as is:
+    read_write_info = ReadWriteInfo()
+    postfix = ExtractNode.determine_postfix(read_write_info)
+    assert postfix == "_post"
+    postfix = ExtractNode.determine_postfix(read_write_info,
+                                            postfix="_new_postfix")
+    assert postfix == "_new_postfix"
+
+    # Clash between input variable and a created output variable:
+    read_write_info = ReadWriteInfo()
+    read_write_info.add_read(Signature("var_post"))
+    read_write_info.add_write(Signature("var"))
+    postfix = ExtractNode.determine_postfix(read_write_info)
+    assert postfix == "_post0"
+
+    # Two clashes between input variable and a created output variable:
+    read_write_info.add_read(Signature("var_post0"))
+    postfix = ExtractNode.determine_postfix(read_write_info)
+    assert postfix == "_post1"
+
+    # Two clashes between different input variables and created output
+    # variables: 'var1' prevents the '_post' to be used, 'var2'
+    # prevents "_post0" to be used, 'var3' prevents "_post1":
+    read_write_info = ReadWriteInfo()
+    read_write_info.add_read(Signature("var1_post"))
+    read_write_info.add_read(Signature("var2_post0"))
+    read_write_info.add_read(Signature("var3_post1"))
+    read_write_info.add_write(Signature("var1"))
+    read_write_info.add_write(Signature("var2"))
+    read_write_info.add_write(Signature("var3"))
+    postfix = ExtractNode.determine_postfix(read_write_info)
+    assert postfix == "_post2"
+
+    # Handle clash between output variables: the first variable will
+    # create "var" and var_post", the second "var_post" and "var_post_post".
+    read_write_info = ReadWriteInfo()
+    read_write_info. add_write(Signature("var"))
+    read_write_info. add_write(Signature("var_post"))
+    postfix = ExtractNode.determine_postfix(read_write_info)
+    assert postfix == "_post0"
+    read_write_info.add_write(Signature("var_post0"))
+    postfix = ExtractNode.determine_postfix(read_write_info)
+    assert postfix == "_post1"
+
+
+@pytest.mark.usefixtures("change_into_tmpdir")
+def test_psylayer_flatten_same_symbols():
+    '''Make sure that when we flatten a symbol, we bring its data back
+    after using the flattened version, including repeatedly if there are
+    multiple extraction regions.
+
+    '''
+    psy, invoke = get_invoke("driver_test.f90", "gocean",
+                             idx=3, dist_mem=False)
+
+    etrans = GOceanExtractTrans()
+    etrans.apply(invoke.schedule[0])
+    etrans.apply(invoke.schedule[1])
+    etrans.apply(invoke.schedule[2])
+    code = psy.gen
+
+    # Flatten before region 1
+    assert """
+    out_fld_internal_ystart = out_fld%internal%ystart
+    out_fld_internal_ystop = out_fld%internal%ystop
+    out_fld_internal_xstart = out_fld%internal%xstart
+    out_fld_internal_xstop = out_fld%internal%xstop
+    out_fld_data = out_fld%data
```

Comment:
I have to admit I didn't notice that before. So any field of a kernel is actually copied before a call, and then copied back? That seems to be very inefficient. Is this only in gocean (since LFRic seems to be using pointers there?)
No need to sort this out here, but potentially if this an be improved, please open a ticket.

Reply:
I misunderstood what you meant. See new issue #3027

In fact I started implementing it and it worked for the extraction part (this test), see #3027 

But then I realised that the driver_creation part that converts pointers to allocatables is currently in the LFRic subclass and has to be moved to the base class and I also want test that this work with generic code. So the changes keept growing and I decided doing it in a separate PR.